lib/gssapi/sanon/context_time.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 1997 - 2003 Kungliga Tekniska Högskolan
+ * (Royal Institute of Technology, Stockholm, Sweden).
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "sanon_locl.h"
+
+OM_uint32
+_gss_sanon_lifetime_left(OM_uint32 *minor_status,
+			 OM_uint32 endtime,
+			 OM_uint32 *lifetime_rec)
+{
+    krb5_timestamp now;
+    krb5_error_code kret;
+    krb5_context context;
+
+    GSSAPI_KRB5_INIT (&context);
+
+    if (endtime == 0) {
+	*lifetime_rec = GSS_C_INDEFINITE;
+	return GSS_S_COMPLETE;
+    }
+
+    kret = krb5_timeofday(context, &now);
+    if (kret) {
+        *lifetime_rec = 0;
+	*minor_status = kret;
+	return GSS_S_FAILURE;
+    }
+
+    if (endtime < now)
+	*lifetime_rec = 0;
+    else
+	*lifetime_rec = endtime - now;
+
+    return GSS_S_COMPLETE;
+}
+
+
+OM_uint32 GSSAPI_CALLCONV
+_gss_sanon_context_time(OM_uint32 *minor,
+			gss_const_ctx_id_t context_handle,
+			OM_uint32 *time_rec)
+{
+    OM_uint32 endtime;
+    OM_uint32 major_status;
+    const sanon_ctx sc = (const sanon_ctx) context_handle;
+    gsskrb5_ctx rfc3961 = (gsskrb5_ctx)sc->rfc3961;
+
+    if (rfc3961 == NULL)
+	return GSS_S_NO_CONTEXT;
+
+    HEIMDAL_MUTEX_lock(&rfc3961->ctx_id_mutex);
+    endtime = rfc3961->endtime;
+    HEIMDAL_MUTEX_unlock(&rfc3961->ctx_id_mutex);
+
+    major_status = _gss_sanon_lifetime_left(minor, endtime, time_rec);
```

Comment:
When would the context time not be `GSS_C_INDEFINITE` anyways?  When the credential was acquire with some specific `time_req`?  But you can get more time...  I'd say simplify and always indicate `GSS_C_INDEFINITE`.